plugin/main/src/kotlinx/benchmark/gradle/AnnotationsValidator.kt

```diff
@@ -0,0 +1,84 @@
+package kotlinx.benchmark.gradle
+
+import kotlinx.benchmark.gradle.SuiteSourceGenerator.Companion.paramAnnotationFQN
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.builtins.UnsignedTypes
+import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.descriptors.PropertyDescriptor
+import org.jetbrains.kotlin.js.descriptorUtils.getKotlinTypeFqName
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.resolve.annotations.argumentValue
+
+internal fun validateBenchmarkFunctions(functions: List<FunctionDescriptor>) {
+    functions.forEach { function ->
+        if (function.visibility != DescriptorVisibilities.PUBLIC) {
+            error("@Benchmark function should be public. Function `${function.name}` is ${function.visibility.name}.")
+        }
+
+        val parameters = function.valueParameters.size
+        if (parameters == 1) {
+            val paramType = function.valueParameters[0].type
+            if (paramType.getKotlinTypeFqName(false) != "kotlinx.benchmark.Blackhole") {
+                error("@Benchmark function can have at most one parameter of type `Blackhole`. " +
+                    "Function `${function.name}` has a parameter of type `$paramType`. ")
+            }
+        } else if (parameters != 0) {
+            error("@Benchmark function can have at most one parameter of type `Blackhole`. " +
+                    "Function `${function.name}` has $parameters parameters.")
+        }
+    }
+}
+
+internal fun validateSetupFunctions(functions: List<FunctionDescriptor>) {
```

Comment:
Note that on JVM, where our annotations are type aliases to JMH ones, one may annotate functions with kotlinx's `Setup`, `Benchmark` and `Teardown`, but then plug JMH-specific parameters into them:

```kotlin
import kotlinx.benchmark.Benchmark
import kotlinx.benchmark.Scope
import kotlinx.benchmark.Setup
import kotlinx.benchmark.State
import org.openjdk.jmh.infra.BenchmarkParams
import org.openjdk.jmh.infra.Control
import org.openjdk.jmh.infra.IterationParams
import org.openjdk.jmh.infra.ThreadParams

@State(Scope.Benchmark) 
public open class SampleBenchmark {
    @Setup 
    public fun setup(bmParams: BenchmarkParams, tParams: ThreadParams, itParams: IterationParams) {
    }

    @Benchmark 
    public fun benchmarkWithControl(ctrl: Control): Boolean = ctrl.stopMeasurement
}
```

All these parameters are a niche feature, but as long as kotlinx-benchmark does not provide alternatives, it would be to allow users to use them.

Here's a more practical example - https://github.com/qurbonzoda/MutableMapEntryBenchmarks/blob/ce674d61b8e9e798fa76c6a926a2129720eac359/src/main/kotlin/PolymorphicMapsBenchmark.kt#L52 ;)

Reply:
Ah, it's only about non-JVM benchmarks. My bad, sorry.